Call /start and /end

graeme-hill opened the pull request:
Fixes battlesnakeio/roadmap#31 and battlesnakeio/roadmap#32

I had to make a lot more changes than expected and refactor some things to keep the code well organized and avoid copy-pasta. I just went ahead and did things the way that seemed reasonable to me but I'm definitely looking for feedback so feel free to let me know if you would have done it differently.

graeme-hill commented:
I'll fix those failed checks tomorrow.

dlsteuer commented on rules/api_client.go:
```diff
@@ -0,0 +1,95 @@
+	netClient := createClient(timeout)
+	postResponse, err := netClient.Post(getURL(s.URL, url), "application/json", buf)
+	if err != nil {
+		log.WithError(err).
```
logrus is a structured logging package, so you could change this to 

```
log.WithError(err).WithFields(log.Fields{
    "url": url,
    "id": s.ID,
}).Error("error POSTing to snake")
```

graeme-hill replied:
Oh that's cool. Will do.

dlsteuer commented on rules/death_cause.go:
```diff
@@ -9,4 +9,6 @@ const (
 	// DeathCauseWallCollision is when a snake runs off the board
 	DeathCauseWallCollision = "wall-collision"
+	// DeathCauseStartFail is when a snake does not properly respond to /start
+	DeathCauseStartFail = "start-fail"
```
I'm not sure about this, technically /start isn't required to run a game (it depends how the snake is coded). @bvanvugt should a snake be removed from the game if start fails to respond?

dlsteuer commented on rules/start.go:
```diff
@@ -0,0 +1,68 @@
+func StartSnakes(game *pb.Game, startState *pb.GameTick) {
+	// Be nice and give snake servers a long time to respond to /start in case
+	// it's a sleeping heroku dyno or something like that.
+	timeout := 15 * time.Second
```
@bvanvugt thoughts on this?

graeme-hill replied:
Ya I was curious what people would think about this. The obvious downside is that if a server is dead you might have to sit around for 15 seconds which could get annoying.

dlsteuer commented on worker/runner.go:
```diff
@@ -19,9 +19,8 @@ func Runner(ctx context.Context, client pb.ControllerClient, id string) error {
 	for {
 		if lastTick != nil && lastTick.Turn == 0 {
-			rules.NotifyGameStart(resp.Game)
+			rules.StartSnakes(resp.Game, lastTick)
```
I think I prefer the name NotifyGameStart to keep it consistent with NotifyGameEnd

graeme-hill replied:
To me Notify means "I'm telling you this thing, but I don't care what you do with it". The way I actually implemented it is "I'm telling you this thing and I require that you respond and tell me what color you are or else I will kill you". If I was wrong that /start is required then yes I think Notify is better and I should also remove `DeathCauseStartFail`.

Now that I think about it, it's probably better not to require /start. There's no reason to be that strict. Does this make sense?

* Don't kill snakes that ignore /start
* Use the word Notify
* If the snake doesn't respond to /start with a color just assign a random color

graeme-hill commented:
Ok the following changes have been made:

* /start is optional (like /end). Snakes never get killed for not implementing start or for sending invalid response.
* Snakes are always assigned a color after /start. If they do not successfully set a color in the response then a random color is assigned.
* `StartSnakes` renamed to `NotifyGameStart`
* Added some more tests and fixed a couple bugs they exposed.

Still waiting for feedback on whether the long timeout for /start is a good idea or not.

dlsteuer commented:
Thanks for all the work on this, I really like the refactoring and effort put in. I think 15 seconds is too long, and would probably use 5 seconds instead. Ultimately I think this is a decision for @bvanvugt as he is currently handling the game master role.

graeme-hill commented:
Five seconds it is. I can always change it if @bvanvugt wants to weigh in.

dlsteuer commented:
Sounds great, let's go with 5s for now, and I've pulled the rules questions out to issues that can be tracked in the roadmap.
